### PR TITLE
Added profiler zero stack check

### PR DESCRIPTION
**Pull request by SmilyOrg**
… and removed the faulty loom_tick block as the dump happens within it

Also tested random START/END calls and the only one that fell through was having a START with no END (stack overflow gets caught, but it's possible to not overflow before dumping/reset), which now gets caught with the zero stack check before dumping.

### CODE REVIEW

File: loom/engine/bindings/loom/lmApplicationTasks.cpp [hunk 1]
```diff
@@ -33,8 +33,6 @@ extern "C"
 
 void loom_tick()
 {
-    LOOM_PROFILE_START(loom_tick);
```
Comment (bengarney) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: Why do we want to remove this in favor of the profilerBlock_t?

File: loom/common/core/performance.cpp [hunk 2]
```diff
@@ -606,6 +618,8 @@ static void LoomProfilerEntryDumpRecurse(LoomProfilerEntry *data, char *buffer,
 
 void LoomProfiler::dump()
 {
+    LOOM_PROFILE_ZERO_CHECK()
```
Comment (bengarney) [inline loom/common/core/performance.cpp]: Is this the actual dump implementation or the thing that requests a dump later?

File: loom/common/core/performance.cpp [hunk 3]
```diff
@@ -526,6 +526,18 @@ void LoomProfiler::hashPop(LoomProfilerRoot *expected)
     }
 }
 
+void LoomProfiler::hashZeroCheck()
```
Comment (bengarney) [inline loom/common/core/performance.cpp]: That's nice

**Comment by bengarney** [comment]
Nitted!

File: loom/engine/bindings/loom/lmApplicationTasks.cpp (hunk 1 above)
Comment (SmilyOrg) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: Not really in favor of, but I don't think it gets tracked accurately anyway, since the block itself contains the profiling tick and when the profile gets dumped, the tick is in progress and therefore fails the zero stack check (we could add an exception, but that's kind-of odd).

File: loom/common/core/performance.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/common/core/performance.cpp]: It just calls `hashZeroCheck` from anywhere and follows the style of `LOOM_PROFILE_START/END`.

**Comment by SmilyOrg** [comment]
Counterpoint!

File: loom/engine/bindings/loom/lmApplicationTasks.cpp (hunk 1 above)
Comment (bengarney) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: So what changes in the profile dump when you remove this block?

**Comment by bengarney** [comment]
Renitted!

File: loom/engine/bindings/loom/lmApplicationTasks.cpp (hunk 1 above)
Comment (SmilyOrg) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: `loom_tick` gets replaced with `ROOT` I think. I'll recheck.

Comment (bengarney) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: dump() should be called the next time the stack goes to zero - it should never be called directly. So that shouldn't happen and is a bug?

Comment (SmilyOrg) [inline loom/engine/bindings/loom/lmApplicationTasks.cpp]: Here's what the new dump looks like:

```
[asset.protocol] LOG: [profiler] Ordered by stack trace total time -
[asset.protocol] LOG: [profiler]   % Time % NSTime  AvgTime  MaxTime  MinTime Invoke # Name
[asset.protocol] LOG: [profiler]  100.000  100.000 3930.367    0.000    1.#IO        0 ROOT
[asset.protocol] LOG: [profiler]   98.601    0.269   39.755   45.165    3.214      309   loom.Application.onInternalFrame
[asset.protocol] LOG: [profiler]   92.211   92.211   37.179   42.855    0.941      309     waitForVSync
[asset.protocol] LOG: [profiler]    4.699    1.129    1.895   11.044    1.090      309     stageRenderDisplayList
[asset.protocol] LOG: [profiler]    3.409    0.028    0.035    9.261    0.000    12051       quadGetVertices
[asset.protocol] LOG: [profiler]    3.381    3.365    0.036    9.258    0.014    11742         quadSubmit
[asset.protocol] LOG: [profiler]    1.092    0.110    0.440    1.747    0.246      309     loom2d.animation.Juggler.advanceTime
[asset.protocol] LOG: [profiler]    0.888    0.038    0.358    1.069    0.194      309       loom.modestmaps.core.TileGrid.advanceTime
[asset.protocol] LOG: [profiler]    0.844    0.576    0.340    1.019    0.185      309         loom.modestmaps.core.TileGrid.pruneTiles
[asset.protocol] LOG: [profiler]    0.262    0.024    0.106    0.312    0.054      309           loom.modestmaps.core.painter.TilePainter.getCacheSize
[asset.protocol] LOG: [profiler]    0.238    0.238    0.096    0.287    0.048      309             loom.modestmaps.core.painter.TileCache.__pget_size
[asset.protocol] LOG: [profiler]    0.172    0.171    0.069    0.293    0.038      309     stageRenderBegin
[asset.protocol] LOG: [profiler]    0.107    0.049    0.043    0.152    0.026      309     loom2d.display.Stage.advanceTime
[asset.protocol] LOG: [profiler]    0.593    0.593    0.239    3.068    0.001      309   garbageCollection
[asset.protocol] LOG: [profiler]    0.423    0.001   26.353   52.665    0.041        2   loom.gameframework.ConsoleCommandManager.process
[asset.protocol] LOG: [profiler]    0.420    0.419   52.340   52.340   52.340        1     loom.Application.onProfilerDump
[asset.protocol] LOG: [profiler]    0.351    0.082    0.142    0.617    0.056      309   loom.gameframework.TimeManager.process
[asset.protocol] LOG: [profiler]    0.270    0.112    0.109    0.518    0.043      309     loom.gameframework.TimeManager.advance
[asset.protocol] LOG: [profiler]    0.144    0.078    0.058    0.447    0.023      309       loom.gameframework.TimeManager.fireTick
[asset.protocol] LOG: [profiler] Suppressed 35 items with < 0.1% of measured time.
```

**Comment by SmilyOrg** [comment]
Countercounterpoint!

**Comment by Newtsuck** [comment]
Looks good to me!

**Comment by bengarney** [comment]
GC'd. Ready to land if you guys want to land it.

**Comment by SmilyOrg** [comment]
Yeah seems fine to me.

**Comment by bengarney** [comment]
OK so land it please. :)
